### ISSUE TITLE
Launchd Incompatibilities

### ISSUE THREAD

**Issue by romw** [BOINC/boinc-ng#759]
**Reported by dak180 on 1 Jun 38902419 20:00 UTC**
With the attached plist launchd restarts boinc continuously because there is already a copy running.

This leads me to believe that boinc is '''not''' following [the rules for a launchd controlled daemon](http://developer.apple.com/DOCUMENTATION/Darwin/Reference/ManPages/man5/launchd.plist.5.html#//apple_ref/doc/man/5/launchd.plist).

```
A daemon or agent launched by launchd MUST NOT do the following
 in the process directly launched by launchd:

  oo   Call daemon(3).
  oo   Do the moral equivalent of daemon(3) by calling fork(2)
        and have the parent process exit(3) or _exit(2).
```

I am running the OS X version which is '''6.2.18''' not '''6.2.19'''.

**Comment by romw** [comment]
**Commented by Nicolas on 21 Sep 38903452 05:46 UTC**
If you use the `-daemon` switch, daemon(3) will be called. So... don't use that switch?

**Comment by romw** [comment]
**Commented by KSMarksPsych on 18 May 38903608 02:13 UTC**
Reassigning this to Charlie as it's a Mac thing.

**Comment by romw** [comment]
**Commented by dak180 on 29 Mar 38904334 20:00 UTC**
Replying to [Nicolas](comment:2):
> If you use the `-daemon` switch, daemon(3) will be called. So... don't use that switch?

I may be wrong, but I do not think that that is all that the '-daemon' switch does.

Even if that were the case the plist file that is included in [14490] has the switch in it and should not.

**Comment by romw** [comment]
**Commented by charlief on 13 Oct 38905016 19:33 UTC**
Replying to [dak180](comment:4):
> Replying to [Nicolas](comment:2):
> > If you use the `-daemon` switch, daemon(3) will be called. So... don't use that switch?
> 
> I may be wrong, but I do not think that that is all that the '-daemon' switch does.
> 
> Even if that were the case the plist file that is included in [14490] has the switch in it and should not.

First, BOINC does follow all the rules in the document referenced by dak180.  It does not call either daemon() or fork() followed by exit() or _exit().  

Second, the -daemon argument does not cause BOINC to call daemon (3).  The affect of arguments is determined solely by how they are used by BOINC Client's code.  The -daemon argument in the plist file is correct and necessary; BOINC uses this to tell it not to exit automatically when it detects that the manager has quit or crashed.  

Third, the problem arises because dak180 has modified the original plist which is installed by the Make_BOINC_Service.sh script available on the BOINC web site.  The original script works correctly.  He has added the KeepAlive tag, which has introduced this problem.

However, the KeepAlive tag should work.  I'm investigating this issue.  It seems that even though launchd does properly launch BOINC at system boot, it fails to recognize that it is running after that and tries to launch it again with the KeepAlive tag.  I've also noticed that `launchctl stop edu.berkeley.boinc` doesn't work properly; I suspect these issues are related.

I'll  update this ticket when I have more information.

**Comment by romw** [comment]
**Commented by charlief on 20 Oct 38906455 02:13 UTC**
(In [16565]) client: Fix error from changeset 14415 which caused Mac client to call daemon() if it was passed -daemon argument, causing problems using launchd with KeepAlive option; fixes #782.
